### PR TITLE
fix(NODE-5064): consolidate connection cleanup logic and ensure socket is always closed

### PR DESCRIPTION
**Pull request by baileympearson**
#### What is changing?

This PR primarily does three things:

- ensures that the message stream is always destroyed and ensure that the socket is always destroyed
- consolidate the connection clean up logic into a single function that is called from each point where the connection is cleaned out
- improves the unit tests to properly assert on the state of the connection after _cleanup has been called. Additionally, the unit tests now in for synchronous versus asynchronous behavior for the various error and clean up scenarios.

Example unit test output:
```
  new Connection()
    ✔ should support fire-and-forget messages
    ✔ should destroy streams which time out (55ms)
    ✔ should throw a network error with kBeforeHandshake set to false on timeout after handshake (56ms)
    ✔ should throw a network error with kBeforeHandshake set to true on timeout before handshake (53ms)
    #onMessage
      when the connection is a monitoring connection
        when multiple hellos exist on the stream
          ✔ calls the callback with the last hello document
        when requestId/responseTo do not match
          ✔ calls the operation description callback with the document
        when requestId/reponseTo match
          ✔ calls the operation description callback with the document
        when no operation description is in the queue
          ✔ does not error
        when more than one operation description is in the queue
          ✔ calls all operation description callbacks with an error
    when the socket times out
      delayed timeout for lambda behavior
        ✔ should delay timeout errors by one tick
        ✔ should clear timeout errors if more data is available
      when the timeout expires and no more data has come in
        ✔ destroys the message stream
        ✔ ends the socket
        ✔ destroys the socket after ending it
        ✔ passes the error along to any callbacks in the operation description queue (asynchronously)
        ✔ emits a Connection.CLOSE event (asynchronously)
        ✔ removes all listeners on the message stream
        ✔ removes all listeners on the socket
    when the message stream errors
      ✔ destroys the message stream synchronously
      ✔ ends the socket
      ✔ destroys the socket after ending it (asynchronously)
      ✔ passes the error along to any callbacks in the operation description queue (asynchronously)
      ✔ emits a Connection.CLOSE event (asynchronously)
      ✔ removes all listeners on the message stream
      ✔ removes all listeners on the socket
    when the underlying socket closes
      ✔ destroys the message stream synchronously
      ✔ ends the socket
      ✔ destroys the socket after ending it (asynchronously)
      ✔ calls any callbacks in the queue with a MongoNetworkError (asynchronously)
      ✔ emits a Connection.CLOSE event (asynchronously)
      ✔ removes all listeners on the message stream
      ✔ removes all listeners on the socket
    .hasSessionSupport
      when logicalSessionTimeoutMinutes is present
        ✔ returns true
      when logicalSessionTimeoutMinutes is not present
        when in load balancing mode
          ✔ returns true
        when not in load balancing mode
          ✔ returns false
    destroy()
      when a callback is provided
        when the connection was already destroyed
          ✔ does not attempt to cleanup the socket again
          ✔ calls the callback (asynchronously)
        when the connection was not destroyed
          ✔ cleans up the connection
          ✔ calls the callback (asynchronously)
      when options.force == true
        ✔ destroys the tcp socket (synchronously)
        ✔ does not call stream.end
        ✔ destroys the messageStream (synchronously)
        ✔ when destroy({ force: true }) is called multiple times, it calls stream.destroy exactly once
      when options.force == false
        ✔ destroys the tcp socket (asynchronously)
        ✔ ends the tcp socket (synchronously)
        ✔ destroys the messageStream (synchronously)
        ✔ calls stream.end exactly once when destroy is called multiple times
```
### Double check the following

- [ ] Ran `npm run check:lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/cmap/connection.ts [hunk 1]
```diff
@@ -293,37 +292,57 @@ export class Connection extends TypedEventEmitter<ConnectionEvents> {
     this[kLastUseTime] = now();
   }
 
-  onError(error: Error) {
+  _cleanup(options: { force: boolean; errorFactory?: () => Error }) {
```
Comment (baileympearson) [inline src/cmap/connection.ts:295]: errorFactory could be changed, this maintains the old behavior each callback receives a different instance of the error. We could also just passed an error indirectly.

File: src/cmap/connection.ts [hunk 2]
```diff
@@ -444,34 +460,13 @@ export class Connection extends TypedEventEmitter<ConnectionEvents> {
...
       return;
     }
-
-    this[kStream].end(() => {
-      this.destroyed = true;
-      if (typeof callback === 'function') {
-        callback();
-      }
-    });
+    this.once('close', () => process.nextTick(() => callback?.()));
```
Comment (baileympearson) [inline src/cmap/connection.ts:467]: I am happy to revert this change. By listening for the close event, the centralized cleanup handler does not need to worry about an optional callback. None of the event handlers rely on it, so it's only necessary in this method.

File: test/unit/cmap/connection.test.ts [hunk 3]
```diff
@@ -362,81 +389,324 @@ describe('new Connection()', function () {
...
 
-        it('calls all operation description callbacks with an error', function () {
+        it('calls all operation description callbacks with an error', function (done) {
           expect(spyOne).to.be.calledOnce;
           expect(spyTwo).to.be.calledOnce;
           const errorOne = spyOne.firstCall.args[0];
           const errorTwo = spyTwo.firstCall.args[0];
           expect(errorOne).to.be.instanceof(MongoRuntimeError);
           expect(errorTwo).to.be.instanceof(MongoRuntimeError);
+          done();
```
Comment (addaleax) [inline test/unit/cmap/connection.test.ts:399]: Is this change intentional?

File: src/cmap/connection.ts (hunk 1 above)
Comment (addaleax) [reply src/cmap/connection.ts:295]: Yeah, it’s not immediately obvious to me why this isn’t `error: Error` instead of `errorFactory`… is it for the stack trace? That would make sense but is probably something that should be noted in a comment

**Review by addaleax — APPROVED** [review]

File: src/cmap/connection.ts (hunk 1 above)
Comment (baileympearson) [reply src/cmap/connection.ts:295]: This technically preserves the existing behavior where each callback receives a distinct instance of the error. I think it's a reasonable argument to change this behavior (and good point about to call stack) so if we're OK passing the same instance to each call back, then i'll make that change.

_Technically_ there should only be one call back in there ever though.

File: src/cmap/connection.ts [hunk 4]
```diff
@@ -293,37 +292,57 @@ export class Connection extends TypedEventEmitter<ConnectionEvents> {
...
+        op.cb(errorFactory?.());
+      }
+
+      this[kQueue].clear();
+
+      this.emit(Connection.CLOSE);
+    };
+
+    this.removeAllListeners(Connection.PINNED);
+    this.removeAllListeners(Connection.UNPINNED);
```
[the remark this answers is not in the document]
Comment (baileympearson) [reply src/cmap/connection.ts:314]: We need to leave the close listener on but that's it. I can remove them all if you'd like

File: src/cmap/connection.ts (hunk 2 above)
Comment (nbbeeken) [reply src/cmap/connection.ts:467]: I was thinking we could make cleanup an async function and we can invoke the callback from a `.then`. If that's reasonable let me know. Otherwise, we can keep this but add a gate: `if (callback) this.once ///...`

File: src/cmap/connection.ts (hunk 1 above)
Comment (addaleax) [reply src/cmap/connection.ts:295]: Right, if object identity for the `Error` objects is the concern here, I’m okay with passing the same `Error` object to multiple callbacks.

File: test/unit/cmap/connection.test.ts [hunk 5]
```diff
@@ -362,81 +389,324 @@ describe('new Connection()', function () {
...
+        cleanupSpy = sinon.spy(connection, '_cleanup');
+        timeoutSpy = sinon.spy(connection, 'onTimeout');
+      });
+
+      afterEach(() => {
+        sinon.restore();
+        timerSandbox.restore();
+        clock.restore();
+        sinon.restore();
+      });
```
Comment (nbbeeken) [inline test/unit/cmap/connection.test.ts:466]: dupe clean up, double sinon.restore and overlap with the above afterEach, I assume this is left from debuggin and not intentional?

File: test/unit/cmap/connection.test.ts [hunk 6]
```diff
@@ -362,81 +389,324 @@ describe('new Connection()', function () {
...
+        expect(messageStream.eventNames()).to.deep.equal([]);
+      });
+
+      it('removes all listeners on the socket', () => {
+        expect(driverSocket.eventNames()).to.deep.equal([]);
+      });
+    });
+  });
+
+  describe('when the message stream errors', () => {
```
Comment (nbbeeken) [inline test/unit/cmap/connection.test.ts:548]:
```suggestion
  describe('when the MessageStream errors', () => {
```

File: test/unit/cmap/connection.test.ts [hunk 7]
```diff
@@ -362,81 +389,324 @@ describe('new Connection()', function () {
...
+      expect(closeCount).to.equal(0);
+      clock.runAll();
+      expect(closeCount).to.equal(1);
+    });
 
-      driverSocket.emit('timeout');
-      expect(connection.onTimeout).to.have.been.calledOnce;
-      expect(connection).to.have.property(kDelayedTimeoutId).that.is.instanceOf(NodeJSTimeoutClass);
+    it('removes all listeners on the message stream', () => {
+      expect(messageStream.eventNames()).to.deep.equal([]);
```
Comment (nbbeeken) [inline test/unit/cmap/connection.test.ts:622]: minor: but listenerCount instead of deep equal empty array?

File: test/unit/cmap/connection.test.ts [hunk 8]
```diff
@@ -362,81 +389,324 @@ describe('new Connection()', function () {
...
-      expect(connection).to.have.property('closed', false);
-      expect(connection).to.have.property(kDelayedTimeoutId, null);
+    it('emits a Connection.CLOSE event (asynchronously)', () => {
+      expect(closeCount).to.equal(0);
+      clock.runAll();
+      expect(closeCount).to.equal(1);
+    });
+
+    it('removes all listeners on the message stream', () => {
+      expect(messageStream.eventNames()).to.deep.equal([]);
```
Comment (nbbeeken) [inline test/unit/cmap/connection.test.ts:705]: listenerCount + assertion message?

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/cmap/connection.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply src/cmap/connection.ts]: Yea, seems like a good call

File: src/cmap/connection.ts [hunk 9]
```diff
@@ -444,34 +413,73 @@ export class Connection extends TypedEventEmitter<ConnectionEvents> {
...
+
+      this.emit(Connection.CLOSE);
+    };
+
+    for (const event of [
+      Connection.MESSAGE,
+      Connection.PINNED,
+      Connection.UNPINNED,
+      Connection.COMMAND_STARTED,
+      Connection.COMMAND_FAILED,
```
Comment (nbbeeken) [inline src/cmap/connection.ts:457]: I think CI is red b/c of this change, commandFailed is emitted when op.cb is called. I think we should move this loop into complete callback after op.cb is called

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/cmap/connection.ts (hunk 9 above)
Comment (nbbeeken) [reply src/cmap/connection.ts:457]: Actually then we can simplify this back to removeAllListeners if we call it after the emit(CLOSE)

**Review by nbbeeken — APPROVED** [review]
